src/webgpu/web_platform/reftests/canvas_complex.html.ts

```diff
@@ -1,78 +1,342 @@
+  output.fragUV = uv[VertexIndex];
+  return output;
+}
+            `,
+          }),
+          entryPoint: 'main',
+        },
+        fragment: {
+          module: t.device.createShaderModule({
+            code: `
+[[group(0), binding(0)]] var mySampler: sampler;
+[[group(0), binding(1)]] var myTexture: texture_2d<f32>;
+
+[[stage(fragment)]]
+fn main([[location(0)]] fragUV: vec2<f32>) -> [[location(0)]] vec4<f32> {
+  return textureSample(myTexture, mySampler, fragUV);
+}
+            `,
+          }),
+          entryPoint: 'main',
+          targets: [
+            {
+              format,
+            },
+          ],
+        },
+        primitive: {
+          topology: 'triangle-list',
+        },
+      });
+
+      const sampler = t.device.createSampler({
+        magFilter: 'nearest',
+        minFilter: 'nearest',
+      });
+
+      const uniformBindGroup = t.device.createBindGroup({
+        layout: pipeline.getBindGroupLayout(0),
+        entries: [
+          {
+            binding: 0,
+            resource: sampler,
+          },
+          {
+            binding: 1,
+            resource: srcTexture.createView(),
+          },
+        ],
+      });
+
+      const renderPassDescriptor: GPURenderPassDescriptor = {
+        colorAttachments: [
+          {
+            view: ctx.getCurrentTexture().createView(),
+
+            loadValue: { r: 0.5, g: 0.5, b: 0.5, a: 1.0 },
+            storeOp: 'store',
+          },
+        ],
+      };
+
+      const commandEncoder = t.device.createCommandEncoder();
+      const passEncoder = commandEncoder.beginRenderPass(renderPassDescriptor);
+      passEncoder.setPipeline(pipeline);
+      passEncoder.setBindGroup(0, uniformBindGroup);
+      passEncoder.draw(6, 1, 0, 0);
+      passEncoder.endPass();
+      t.device.queue.submit([commandEncoder.finish()]);
     }
-    buffer.unmap();
 
-    const texture = swapChain.getCurrentTexture();
+    function DrawVertexColor() {
+      assert(ctx !== null);
+      const pipeline = t.device.createRenderPipeline({
+        vertex: {
+          module: t.device.createShaderModule({
+            code: `
+struct VertexOutput {
+  [[builtin(position)]] Position : vec4<f32>;
+  [[location(0)]] fragColor : vec4<f32>;
+};
+
+[[stage(vertex)]]
+fn main([[builtin(vertex_index)]] VertexIndex : u32) -> VertexOutput {
+  var pos = array<vec2<f32>, 6>(
+      vec2<f32>( 0.5,  0.5),
+      vec2<f32>( 0.5, -0.5),
+      vec2<f32>(-0.5, -0.5),
+      vec2<f32>( 0.5,  0.5),
+      vec2<f32>(-0.5, -0.5),
+      vec2<f32>(-0.5,  0.5));
+
+  var offset = array<vec2<f32>, 4>(
+    vec2<f32>( -0.5,  0.5),
+    vec2<f32>( 0.5, 0.5),
+    vec2<f32>(-0.5, -0.5),
+    vec2<f32>( 0.5,  -0.5));
+
+  var color = array<vec4<f32>, 4>(
+      vec4<f32>(0.49804, 0.0, 0.0, 1.0),
+      vec4<f32>(0.0, 0.49804, 0.0, 1.0),
+      vec4<f32>(0.0, 0.0, 0.49804, 1.0),
+      vec4<f32>(0.49804, 0.49804, 0.0, 1.0)); // 0.49804 -> 0x7f
```

Comment:
FYI for the srgb cases: you can use the existing `gammaDecompress` helper to find the corresponding float value for srgb 0x7f: `0.21223`

(may need to be bumped up or down a bit? I don't know what the rounding behavior for unorm storage is)